File: 1/1.py
def newton(f, Df, x0, epsilon, max_iter):
    '''Approximate solution of f(x)=0 by Newton's method.

    Parameters
    ----------
    f : function
        Function for which we are searching for a solution f(x)=0.
    Df : function
        Derivative of f(x).
    x0 : number
        Initial guess for a solution f(x)=0.
    epsilon : number
        Stopping criteria is abs(f(x)) < epsilon.
    max_iter : integer
        Maximum number of iterations of Newton's method.

    Returns
    -------
    xn : number
        Implement Newton's method: compute the linear approximation
        of f(x) at xn and find x intercept by the formula
            x = xn - f(xn)/Df(xn)
        Continue until abs(f(xn)) < epsilon and return xn.
        If Df(xn) == 0, return None. If the number of iterations
        exceeds max_iter, then return None.

    Examples
    --------
    > f = lambda x: x**2 - x - 1
    > Df = lambda x: 2*x - 1
    > newton(f,Df,1,1e-8,10)
    Found solution after 5 iterations.
    1.618033988749989
    '''

    #########################

    xn = x0
    for n in range(0, max_iter):
        fxn = f(xn)

        if abs(fxn) < epsilon:
            print('Found solution after', n, 'iterations.')
            # return xn
        Dfxn = Df(xn)
        if Dfxn == 0:
            print('Zero derivative. No solution found.')
            return None
        print(fxn, xn, Dfxn)
        xn = xn - fxn / Dfxn
        print(xn)

    print('Exceeded maximum iterations. No solution found.')
    return None


from math import sin, cos
p = lambda x: 2 * x ** 2 + 2 * cos(x)
Dp = lambda x: 4 * x - 2 * sin(x)

# [-2, 0] ==> -1
x0 = -1
epsilon = 1e-10
max_iter = 100
newton(p, Dp, x0, epsilon, max_iter)


# p = lambda x: x**3 - x**2 - 1
# Dp = lambda x: 3*x**2 - 2*x
# print(newton(p,Dp,1,1e-10,10))

File: 1/1.f90
PROGRAM main
    IMPLICIT NONE
    REAL(kind = 8):: x0, eps,f,Df
    INTEGER :: max_iter
    CHARACTER(len=128) :: newton

    x0 = -1
    eps = 0.0000000001
    max_iter = 20

    PRINT *, newton(x0, eps, max_iter)

END PROGRAM

CHARACTER(len=128) FUNCTION newton(x0, eps, max_iter) RESULT(RESULTs)
    REAL(kind = 8):: f,Df,eps,x0,fxn,xn
    INTEGER :: max_iter
    CHARACTER(len=36) :: aString, bString

    RESULTs = "nothing"

    xn = x0
    DO i=0,max_iter-1
        fxn = f(xn)

        IF (ABS(fxn) < eps) then
            WRITE(aString,*) i
            WRITE(bString, *) xn
            RESULTs = "Found solution after" // aString // "iterations." // "RESULT : " // bString
            RETURN
        END IF

        Dfxn = Df(xn)
        IF (Dfxn == 0) then
            RESULTs =  "Zero derivative. No solution found."
            RETURN
        END IF
        xn = xn - (fxn / Dfxn)
    END DO

    RESULTs =  "Exceeded maximum iterations. No solution found."
    RETURN

END FUNCTION

REAL(kind = 8) FUNCTION f(x)
    REAL(kind = 8):: x
    f=2 * x ** 2 + 2 * cos(x)
END FUNCTION

REAL(kind = 8) FUNCTION Df(x)
    REAL(kind = 8):: x
    f=4 * x - 2 * sin(x)
END FUNCTION

!REAL(kind = 8) FUNCTION f(x)
!    REAL(kind = 8):: x
!    f=x**3 - x**2 - 1
!END FUNCTION
!
!REAL(kind = 8) FUNCTION Df(x)
!    REAL(kind = 8):: x
!    f=3*x**2 - 2*x
!END FUNCTION